hr_timesheet_holiday/models/hr_holidays.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Sunflower IT <http://sunflowerweb.nl>
+# Copyright 2019 Coop IT Easy SCRLfs
+#   - Vincent Van Rossem <vincent@coopiteasy.be>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import datetime
+from datetime import timedelta
+
+
+from openerp import models, fields, api, _
+from openerp.exceptions import Warning as UserError
+
+
+class HrHolidays(models.Model):
+    """Update analytic lines on status change of Leave Request"""
+
+    _inherit = "hr.holidays"
+
+    analytic_line_ids = fields.One2many(
+        "account.analytic.line", "leave_id", "Analytic Lines"
+    )
+
+    @api.multi
+    def add_timesheet_line(self, description, date, hours, account):
+        """Add a timesheet line for this leave"""
+        self.ensure_one()
+        # User exists because already check during the holidays_validate
+        user = self.employee_id.user_id
+        self.sudo().with_context(force_write=True).write(
+            {
+                "analytic_line_ids": [
+                    (
+                        0,
+                        False,
+                        {
+                            "name": description,
+                            "date": date,
+                            "unit_amount": hours,
+                            "company_id": self.employee_id.company_id.id,
+                            "account_id": account.id,
+                            "user_id": user.id,
+                            "is_timesheet": True,
+                        },
+                    )
+                ]
+            }
+        )
+
+    @api.model
+    def _get_company_hours_per_day(self, company):
+        hours_per_day = company.timesheet_hours_per_day
+        if not hours_per_day:
+            raise UserError(
+                _("No hours per day defined for Company '%s'")
+                % (company.name,)
+            )
+        return hours_per_day
+
+    @api.model
+    def _get_contract_hours_per_day(self, employee, date):
+        self.ensure_one()
+        start_dt = date
+        end_dt = False
+        from_dt = self._compute_datetime(self.date_from)
+        to_dt = self._compute_datetime(self.date_to)
+        if date.date() == from_dt.date():
+            start_dt = from_dt
+        elif date.date() == to_dt.date():
+            end_dt = to_dt
+        hours_per_day = 0.0
+        contracts = (
+            self.env["hr.contract"]
+            .sudo()
```

Comment:
be careful that when assigning a variable through a `sudo`,  the new variable (here `contracts`) keeps the sudo rights. You could write anything or delete the record here.

Here, you're only reading so it's fine.

Reply:
noted :) thx